hare/haretypes.go

```diff
@@ -30,73 +31,96 @@ func NewBytes32(buff []byte) Bytes32 {
 	return x
 }
 
-func (b32 *Bytes32) Id() uint32 {
+func (b32 Bytes32) Id() uint32 {
 	h := fnv.New32()
 	h.Write(b32[:])
 	return h.Sum32()
 }
 
-func (b32 *Bytes32) Bytes() []byte {
+func (b32 Bytes32) Bytes() []byte {
 	return b32[:]
 }
 
 type Set struct {
-	blocks []BlockId
+	blocks map[uint32]BlockId
 }
 
-func NewSet(data [][]byte) Set {
-	s := Set{}
+func NewEmptySet() *Set {
+	s := &Set{}
+	s.blocks = make(map[uint32]BlockId, layerSize)
 
-	s.blocks = make([]BlockId, len(data))
+	return s
+}
+
+func NewSet(data [][]byte) *Set {
+	s := &Set{}
+
+	s.blocks = make(map[uint32]BlockId, len(data))
 	for i := 0; i < len(data); i++ {
-		copy(s.blocks[i].Bytes(), data[i])
+		bid := BlockId{NewBytes32(data[i])}
+		s.blocks[bid.Id()] = bid
 	}
 
 	return s
 }
 
-func (s *Set) To2DSlice() [][]byte {
-	slice := make([][]byte, len(s.blocks))
-	for i := 0; i < len(s.blocks); i++ {
-		copy(slice[i], s.blocks[i].Bytes())
+func (s *Set) Contains(id BlockId) bool {
+	_, exist := s.blocks[id.Id()]
+	return exist
+}
+
+func (s *Set) Add(id BlockId) {
+	if _, exist := s.blocks[id.Id()]; exist {
+		return
 	}
 
-	return slice
+	s.blocks[id.Id()] = id
 }
 
-type AggregatedMessages struct {
-	messages []*pb.HareMessage
-	aggSig   Signature
+func (s *Set) Remove(id BlockId) {
+	delete(s.blocks, id.Id())
 }
 
-func NewAggregatedMessages(p *pb.AggregatedMessages) AggregatedMessages {
-	m := AggregatedMessages{}
+func (s *Set) Equals(g *Set) bool {
+	if len(s.blocks) != len(g.blocks) {
+		return false
+	}
 
-	m.messages = make([]*pb.HareMessage, len(p.Messages))
-	for i := 0; i < len(p.Messages); i++ {
-		m.messages[i] = p.Messages[i]
+	for _, bid := range s.blocks {
+		if _, exist := g.blocks[bid.Id()]; !exist {
+			return false
+		}
 	}
-	m.aggSig = make([]byte, len(p.AggSig))
-	copy(m.aggSig, p.AggSig)
 
-	return m
+	return true
 }
 
-type Certificate struct {
-	set Set
-	AggregatedMessages
+func (s *Set) To2DSlice() [][]byte {
+	slice := make([][]byte, len(s.blocks))
+	i := 0
+	for _, v := range s.blocks {
+		slice[i] = make([]byte, 32)
```

Comment:
The value of 32 should be a const var (so we can easily change it later)

Reply:
Actually, it should even be the len of the source. Fixed.